### PR TITLE
refactor of the s3 input

### PR DESCRIPTION
**Pull request by ph**
This PRs includes a lot of changes.

Ruby code cleanup, following the STYLE.md
It now use the aws_config mixing support IAM roles fix #1778 fix #1575
the plugin also use standard codec
refactoring of the code to improve clarity and allow easier testing
Created initial test suite for it.
Cloudfront is now a codec fix #1630
Gzip decode is now a codec initial work for #1895

### CODE REVIEW

**Comment by geez** [comment]
Hi fellas,

Any idea when this will be merged?

**Comment by ph** [comment]
@purbon since you are reviewing the s3 output can you review this one too?

File: logstash-input-s3.gemspec [hunk 1]
```diff
@@ -22,8 +22,10 @@ Gem::Specification.new do |s|
   # Gem dependencies
   s.add_runtime_dependency 'logstash', '>= 1.4.0', '< 2.0.0'
   s.add_runtime_dependency 'logstash-mixin-aws'
+  s.add_runtime_dependency 'logstash-codec-line'
   s.add_runtime_dependency 'logstash-codec-plain'
```
Comment (purbon) [inline logstash-input-s3.gemspec]: We can remove this as we're not using it anymore.

File: spec/inputs/s3_spec.rb [hunk 2]
```diff
@@ -1,5 +1,147 @@
...
+    it 'should allow user to exclude files from the s3 bucket' do
+      allow_any_instance_of(AWS::S3::ObjectCollection).to receive(:with_prefix).with(nil) { objects_list }
+
+      config = LogStash::Inputs::S3.new(settings.merge({ "exclude_pattern" => "^exclude" }))
+      config.register
+      expect(config.list_new_files).to eq([present_object.key])
+    end
+
+    it 'should support not providing a exclude pattern' do
+      allow_any_instance_of(AWS::S3::ObjectCollection).to receive(:with_prefix).with(nil) { objects_list }
```
Comment (purbon) [inline spec/inputs/s3_spec.rb]: This should be extracted into a before hook

File: lib/logstash/inputs/s3.rb [hunk 3]
```diff
@@ -60,19 +62,55 @@ class LogStash::Inputs::S3 < LogStash::Inputs::Base
...
     require "digest/md5"
     require "aws-sdk"
 
-    @region_endpoint = @region if @region && !@region.empty?
+    @region = get_region
+
+    @logger.info("Registering s3 input", :bucket => @bucket, :region => @region)
+
+    if @sincedb_path.nil?
+      sincedb_file = File.join(ENV["HOME"], ".sincedb_" + Digest::MD5.hexdigest("#{@bucket}+#{@prefix}"))
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: consmetic comment, can we make this a method? it might be more clean.

File: lib/logstash/inputs/s3.rb [hunk 4]
```diff
@@ -60,19 +62,55 @@ class LogStash::Inputs::S3 < LogStash::Inputs::Base
...
+      end
+    end
+
+    unless @backup_to_dir.nil?
+      Dir.mkdir(@backup_to_dir, 0700) unless File.exists?(@backup_to_dir)
+    end
+  end # def register
 
-    @logger.info("Registering s3 input", :bucket => @bucket, :region_endpoint => @region_endpoint)
+  def get_region
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: this methods might be private, and for this go at the end of the file.

File: lib/logstash/inputs/s3.rb [hunk 5]
```diff
@@ -60,19 +62,55 @@ class LogStash::Inputs::S3 < LogStash::Inputs::Base
...
+    else
+      @region
+    end
+  end
 
-    if @credentials.length == 0
-      @access_key_id = ENV['AWS_ACCESS_KEY_ID']
-      @secret_access_key = ENV['AWS_SECRET_ACCESS_KEY']
-    elsif @credentials.length == 1
+  def get_s3object
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: same here, this should be private.

File: lib/logstash/inputs/s3.rb [hunk 6]
```diff
@@ -91,189 +129,165 @@ def register
...
+        :secret_access_key => @secret_access_key,
+        :region => @region
+      )
+    else
+      s3 = AWS::S3.new(aws_options_hash)
     end
+  end
 
-  end # def register
+  public
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: why is this method public?

File: lib/logstash/inputs/s3.rb [hunk 7]
```diff
@@ -91,189 +129,165 @@ def register
...
 
   public
   def run(queue)
-    loop do
-      process_new(queue)
-      sleep(@interval)
+    Stud.interval(@interval) do
+      process_files(queue)
     end
-    finished
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: Think this was an error, why this was used before? notification when the process was finished?

File: lib/logstash/inputs/s3.rb [hunk 8]
```diff
@@ -91,189 +129,165 @@ def register
...
+      unless ignore_filename?(log.key)
+
+        if @sincedb.newer?(log.last_modified)
+          objects[log.key] = log.last_modified
+          @logger.debug("S3 input: Adding to objects[]", :key => log.key)
+        end
       end
     end
-
     return sorted_objects = objects.keys.sort {|a,b| objects[a] <=> objects[b]}
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: why do we need to sort the files?

File: lib/logstash/inputs/s3.rb [hunk 9]
```diff
@@ -91,189 +129,165 @@ def register
...
+    end
+  end
 
-  private
-  def sincedb_read()
+  public
+  def backup_to_bucket(object, key)
+    unless @backup_to_bucket.nil?
+      backup_key = "#{@backup_add_prefix}#{key}"
+      if @delete
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: the name of this variable might be a bit misleading, as the file gets move and not deleted.

File: lib/logstash/inputs/s3.rb [hunk 10]
```diff
@@ -91,189 +129,165 @@ def register
...
+        s3file.write(chunk)
       end
     end
-    return metadata
+  end
 
-  end # def process_line
+  def delete_file_from_bucket
+    if @delete and @backup_to_bucket.nil?
+      object.delete()
```
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: not going to work, object variable not available (might be missing parameter?)

**Comment by purbon** [comment]
As a general comment, I would rather keep in mind what method should be public and which not, I don't really see a constant way of doing that.

File: lib/logstash/inputs/s3.rb (hunk 3 above)
Comment (ph) [inline lib/logstash/inputs/s3.rb]: +1

File: lib/logstash/inputs/s3.rb (hunk 9 above)
Comment (ph) [inline lib/logstash/inputs/s3.rb]: Agree, but this is part of the configuration of the plugin, the move actually `deleted` it from the 'source' bucket.
It just more optimized than doing copy/delete.
```
# Whether to delete processed files from the original bucket.
```
https://github.com/logstash-plugins/logstash-input-s3/blob/master/lib/logstash/inputs/s3.rb#L57

File: lib/logstash/inputs/s3.rb (hunk 10 above)
Comment (ph) [inline lib/logstash/inputs/s3.rb]: fixed.

File: lib/logstash/inputs/s3.rb (hunk 8 above)
Comment (ph) [inline lib/logstash/inputs/s3.rb]: AWS will return the files (keys?) in alphabetical order and we have to sort them by `last_modification`. This allow us to keep track what was the last file we have processed.

File: logstash-input-s3.gemspec (hunk 1 above)
Comment (ph) [inline logstash-input-s3.gemspec]: +1

**Comment by ph** [comment]
@purbon I applied a few of the changes you mentioned, I also regrouped the public and the privates method together.

File: lib/logstash/inputs/s3.rb (hunk 7 above)
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: @ph what about that? an error? makes sense to be removed?

File: lib/logstash/inputs/s3.rb (hunk 9 above)
Comment (purbon) [inline lib/logstash/inputs/s3.rb]: ok, I see.

**Comment by purbon** [comment]
LGTM

File: lib/logstash/inputs/s3.rb (hunk 7 above)
Comment (ph) [inline lib/logstash/inputs/s3.rb]: The `finished` method was never executed, the only way to get out of the loop is to raise an exception and the exception is catched higher in the pipeline.

We have to revisit how we handle exception in the pipeline.

**Pull request closed without merging**